grpc/interceptors.go

```diff
@@ -421,3 +454,93 @@ func (dd deadlineDetails) Error() string {
+	// to allowed client certificate SANs (e.g. "ra.boulder") which are allowed to
+	// make RPCs to that service. The set of client names is implemented as a map
+	// of names to empty structs for easy lookup.
+	serviceClientNames map[string]map[string]struct{}
+}
+
+// newServiceAuthChecker takes a GRPCServerConfig and uses its Service stanzas
+// to construct a serviceAuthChecker which enforces the service/client mappings
+// contained in the config.
+func newServiceAuthChecker(c *cmd.GRPCServerConfig) *authInterceptor {
+	names := make(map[string]map[string]struct{})
+	for serviceName, service := range c.Services {
+		names[serviceName] = make(map[string]struct{})
+		for _, clientName := range service.ClientNames {
+			names[serviceName][clientName] = struct{}{}
+		}
+	}
+	return &authInterceptor{names}
+}
+
+// Unary is a gRPC unary interceptor.
+func (ac *authInterceptor) Unary(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+	err := ac.checkContextAuth(ctx, info.FullMethod)
+	if err != nil {
+		return nil, err
+	}
+	return handler(ctx, req)
+}
+
+// Stream is a gRPC stream interceptor.
+func (ac *authInterceptor) Stream(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+	err := ac.checkContextAuth(ss.Context(), info.FullMethod)
+	if err != nil {
+		return err
+	}
+	return handler(srv, ss)
+}
+
+// checkContextAuth does most of the heavy lifting. It extracts TLS information
+// from the incoming context, gets the set of DNS names contained in the client
+// mTLS cert, and returns nil if at least one of those names appears in the set
+// of allowed client names for given service (or if the set of allowed client
+// names is empty).
+func (ac *authInterceptor) checkContextAuth(ctx context.Context, fullMethod string) error {
+	serviceName, _ := splitMethodName(fullMethod)
+
+	allowedClientNames, ok := ac.serviceClientNames[serviceName]
+	if !ok || len(allowedClientNames) == 0 {
+		return fmt.Errorf("service %q has no allowed client names", serviceName)
+	}
+
+	p, ok := peer.FromContext(ctx)
+	if !ok {
+		return fmt.Errorf("unable to fetch peer info from grpc context")
+	}
+
+	if p.AuthInfo == nil {
+		return fmt.Errorf("grpc connection appears to be plaintext")
+	}
+
+	tlsAuth, ok := p.AuthInfo.(credentials.TLSInfo)
+	if !ok {
+		return fmt.Errorf("connection is not TLS authed")
+	}
+
+	if len(tlsAuth.State.VerifiedChains) == 0 || len(tlsAuth.State.VerifiedChains[0]) == 0 {
+		return fmt.Errorf("connection auth not verified")
+	}
+
+	cert := tlsAuth.State.VerifiedChains[0][0]
+
+	for _, clientName := range cert.DNSNames {
+		_, ok := allowedClientNames[clientName]
+		if ok {
+			return nil
+		}
+	}
+
+	return fmt.Errorf(
+		"client names %v are not authorized for service %q (%v)",
+		cert.DNSNames, serviceName, allowedClientNames)
+}
+
+var _ serverInterceptor = (*authInterceptor)(nil)
```

Comment:
I believe this and the similar declarations in this file are here to provide a static compile time interface check? Given that this syntax is rarely used, would you mind leaving a comment here and above to that effect?